internal/pkg/gateway/api.go

```diff
@@ -427,47 +427,94 @@ func (gs *Server) Submit(ctx context.Context, request *gp.SubmitRequest) (*gp.Su
 		return nil, status.Errorf(codes.Unavailable, "no orderer nodes available")
 	}
 
-	// try each orderer in random order
 	var errDetails []proto.Message
-	for _, index := range rand.Perm(len(orderers)) {
-		orderer := orderers[index]
-		logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", orderer.logAddress)
-
-		var response *ab.BroadcastResponse
-		var err error
-		done := make(chan struct{})
-		go func() {
-			defer close(done)
-			ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
-			defer cancel()
-
-			response, err = gs.broadcast(ctx, orderer, txn)
-		}()
-		select {
-		case <-done:
-			// Broadcast completed normally
-		case <-ctx.Done():
-			// Overall submit timeout expired
-			logger.Warn("Submit call timed out while broadcasting to ordering service")
-			return nil, newRpcError(codes.DeadlineExceeded, "submit timeout expired while broadcasting to ordering service")
+	if gs.getChannelConfig(request.ChannelId).ChannelConfig().Capabilities().ConsensusTypeBFT() {
+		// For BFT, need to send transaction to ALL orderers
+		waitCh := make(chan *gp.ErrorDetail, len(orderers))
+		for _, o := range orderers {
+			go func(ord *orderer) {
+				logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress)
+				ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
+				defer cancel()
+				response, err := gs.broadcast(ctx, ord, txn)
+
+				if err != nil {
+					logger.Warnw("Error sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress, "err", err)
+					waitCh <- errorDetail(ord.endpointConfig, err.Error())
+				} else if status := response.GetStatus(); status == common.Status_SUCCESS {
+					waitCh <- nil
+				} else {
+					logger.Warnw("Unsuccessful response sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress, "status", status, "info", response.GetInfo())
+					waitCh <- errorDetail(ord.endpointConfig, fmt.Sprintf("received unsuccessful response from orderer: %s", common.Status_name[int32(status)]))
+				}
+			}(o)
 		}
-
-		if err != nil {
-			errDetails = append(errDetails, errorDetail(orderer.endpointConfig, err.Error()))
-			logger.Warnw("Error sending transaction to orderer", "txID", request.TransactionId, "endpoint", orderer.logAddress, "err", err)
-			continue
+		atLeastOneSuccess := false
+	loop:
+		for i := 0; i < len(orderers); i++ {
```

Comment:
You don't have to wait for all the nodes to respond.
You can wait for 2f+1 responses - nil and exit the function. 
The rest of the goroutines will timeout and the waitCh channel will be deleted

Reply:
Good point. He can return early once he collected 2f+1 positive responses.

Reply:
Agreed.  And conversely it can fail fast if too many failures are received to attain a quorum

Reply:
the other way around does not work
in the current implementation, one order available to the client may be sufficient

Reply:
After discussing with Yacov, I changed the implementation to require 2f+1 successful responses